### ISSUE TITLE
create webpack plugin

### ISSUE THREAD

**Issue by dlech** [pyodide/pyodide#3089]
> webpack plugin

I would appreciate it if someone else could help with this though. I've tried to write a webpack plugin several times but always run into the problem that I was unable to use webpack with Pyodide at all. Perhaps with @mneil's example I would have an easier time, but the webpack plugin API is also very confusing. If it's just a wrapper around `CopyPlugin` then that probably isn't too hard. 

In @mneil's example, he also patches `pyodide.js` to remove the global assignment to `loadPyodide`, but if you use `pyodide.mjs` instead of `pyodide.js` it doesn't assign `loadPyodide` to the global scope. We make `pyodide.js` out of `pyodide.umd.ts` which just adds the `globalThis` assignment to `pyodide.ts`:
https://github.com/pyodide/pyodide/blob/main/src/js/pyodide.umd.ts
Whereas `pyodide.mjs` is made from `pyodide.ts` directly. Though deleting that line is also an option if using webpack to load a module is not possible.

_Originally posted by @hoodmane in https://github.com/pyodide/pyodide/issues/3085#issuecomment-1241155294_

**Comment by dlech** [comment]
Currently using Pyodide with Webpack is possible by doing something like this: https://github.com/mneil/pyodide-webpack

#3085 proposed adding the exports as in https://github.com/mneil/pyodide-webpack/blob/main/patch.js to the Pyodide JavaScript package, but is was suggested that this could be fragile as the exported files could be renamed/moved/deleted. So it was suggested that we create a webpack plugin that knows about Pyodide internals so that users don't have to deal with such breaking changes.

**Comment by dlech** [comment]
Alternate suggestion:

This is subject to the possible breaking changes pointed out above, but as an alternative to using the webpack copy plugin to copy the binary blob files, if we allowed passing individual URLs to `loadPyodide`, then users could write something like this (the `lockFileURL` line is already possible in Pyodide 0.21.x):

```typescript
    const pyodide = await loadPyodide({
        indexURL: {
            distutilsURL: new URL('pyodide/distutils.tar', import.meta.url).toString(),
            pyodidePyURL: new URL('pyodide/pyodide_py.tar', import.meta.url).toString(),
            pyodideAsmDataURL: new URL('pyodide/pyodide.asm.data, import.meta.url).toString(),
            pyodideAsmJsURL: new URL('pyodide/pyodide.asm.js', import.meta.url).toString(),
            pyodideAsmWasmURL: new URL('pyodide/pyodide.asm.wasm'', import.meta.url).toString(),
        },
        lockFileURL: new URL('pyodide/repodata.json', import.meta.url).toString(),
    });
``` 
(If the API allowed the `URL` type in addition to `string`, we could leave out the `toString()`).

This would probably event work with Create React App or similar tools that don't allow access to the webpack config since it doesn't require any special config.

**Comment by dlech** [comment]
If we go with the copy plugin route, one thing to watch out for is that we don't want different versions of files with the same name.

For example, if we simply do:
```js
      new CopyPlugin({
        patterns: [
          { from: require.resolve("pyodide/distutils.tar"), to: "pyodide/distutils.tar" },
          { from: require.resolve("pyodide/repodata.json"), to: "pyodide/repodata.json" },
          { from: require.resolve("pyodide/pyodide_py.tar"), to: "pyodide/pyodide_py.tar" },
          { from: require.resolve("pyodide/pyodide.asm.data"), to: "pyodide/pyodide.asm.data" },
          { from: require.resolve("pyodide/pyodide.asm.js"), to: "pyodide/pyodide.asm.js" },
          { from: require.resolve("pyodide/pyodide.asm.wasm"), to: "pyodide/pyodide.asm.wasm" },
        ],
      }),
```

Then when upgrading the `pyodide` package and publishing an app, the new versions of these files will have the same name as the old ones and so the web browser may fetch cached versions of these files from a previous version, which of course would likely cause problems.

One way around this is to somehow incorporate the version into the file paths. Some ideas about this have already been discussed in #3068.

**Comment by mneil** [comment]
The purpose of copying these files is so that the webpage loading pyodide knows where to find them. Emscripten does a bunch of implicit loading aided by the rollup.js config that pyodide is using that requires these files to be available at `/pyodide/*`. 

Regarding them being part of the exports... they really should be. I patched the package as part of webpack because otherwise these 6 require.resolve calls fail. Because pyodide has an exports mapping already and it does not explicitly mark these files as exported node marks them as private. The only other way to handle this would be to not use require.resolve and instead use the path module - but that gets messier. What happens when you move these files in the package? require.resolve will find them from exports.

I think it's incorrect to assume these files are private and will not be imported. Obviously you have to have all of these files for pyodide to actually work. Marking them private by excluding them from exports only makes using it in novel ways harder. Other options are using express static routes to resolve to the pyodide folder but that also runs into resolving what is exported and what is not when trying to set up the static route.

For versioning this can be done through query string busting. Using the query string to bust cache has been common for at least the past 15 years that I've done web development. Putting the version in the path makes finding the latest version more difficult. This is a problem that needs to be resolved on the end user side. Cache busting it not a concern of a package maintainer IMO.

In short:

- We absolutely should export these files in the package.json. Marking them private makes using them in novel ways difficult and you're protecting nothing. I can simply copy these files by hand.
- Cache busting is the job of the end-user / consumer and can be handled without us trying to put things into versioned paths

**Comment by mneil** [comment]
Also, FWIW, I have pyodide working in a still private React repository that I will share in the future.

**Comment by dlech** [comment]
For CI, I think it could be useful to have a separate repo. This would allow for additional config to be added for regression tests, i.e. for #3086 and #3087, but we would not want this extra config in the minimal "webpack howto" example.

**Comment by rth** [comment]
You mean a separate repo for the plugin (in addition to the webpack example repo)? Sure, that would work. I think the easiest is that you or @mneil start one, and once it works we can move it to the pyodide org if you want. A separate repo would be easier (rather than including it in Pyodide) as the CI in the monorepo currently takes a long time. Though we can still additionally run this plugin in the main CI, to make sure we don't break anything.

**Comment by mneil** [comment]
I've scaffolded a plugin for webpack that is currently just the example from webpack docs. But I did take a stab at all the supporting docs around it as well as ensuring we can write the plugin in typescript and output a module for use in commonjs and module based projects for webpack.

https://github.com/mneil/pyodide-webpack-plugin

**Comment by rth** [comment]
Great, thank you!

**Comment by mneil** [comment]
Updated the plugin to be what I think is feature complete. It's missing any meaningful tests although I can confirm it works and have been testing it against https://github.com/mneil/pyodide-webpack/tree/index-url. I have not published it to npm yet. I didn't know if pyodide wanted to take it in "officially" and handle releases or if I should just keep it in my personal organization and release it myself. Open to whatever. Let me know if there's something you think it's missing, is incorrect (assumptions on licenses MPL-2.0, contribution guides, etc...) or you'd like to see.

https://github.com/mneil/pyodide-webpack-plugin

**Comment by rth** [comment]
If you feel like you would be interested in maintaining it in the future I would be +1 for accepting it within the Pyodide org with you as its maintainer (if you are interested).

No matter where it would end up you should be able to do the release independently, I think. About npm release the only thing I'm not sure about is whether it would be better to publish it as `pyodide-webpack` or `@pyodide/pyodide-webpack` or some other variation (e.g. `@pyodide/webpack`). Not really sure about best practices with npm scopes.

WDYT @hoodmane @ryanking13 ?

**Comment by ryanking13** [comment]
> If you feel like you would be interested in maintaining it in the future I would be +1 for accepting it within the Pyodide org with you as its maintainer (if you are interested).

+1 if mneil has interests in maintaining the project in the future.

**Comment by mneil** [comment]
I plan to maintain the plugin as I have a personal interest in using it and in wasm / pyodide. I can commit some time each week to answering questions and providing feedback / updates as needed. I imagine most core things will be small (webpack version changes are infrequent, pyodide itself shouldn't have too many major changes to the distribution unless it's planned or emscripten changes). Time would likely be most needed in community feedback.

Anyway, I can publish and maintain it myself under pyodide-webpack-plugin. If you want to pull it into the pyodide org and grant me access to maintain it there I will and we should publish it under @pyodide/webpack-plugin or similar. On that node pyodide really should be published under @pyodide/pyodide. The @pyodide npm organization would help signify that an org (and not an individual) owns the package(s) which helps signal stability. That's a separate issue though :)

**Comment by rth** [comment]
Sounds good. I don't think there will be much maintenance indeed. 

I sent you an invite to the GH org. Could you please transfer the repo to me and I'll transfer it back to the Pyodide org, as I don't think you would have access rights currently to create new repos there.

About where to publish it on npm, yes, `@pyodide/webpack-plugin` sounds good. I added you to the webpack team there, try make the release once the repo is transferred (and if it doesn't work let me know).

As what to do with `@pyodide/pyodide`  indeed, maybe we could discuss that in a separate issue.

**Comment by mneil** [comment]
Transferred the repository to you @rth.

**Comment by rth** [comment]
OK, moved it to https://github.com/pyodide/pyodide-webpack-plugin. It would be nice to switch to this plugin in  https://github.com/pyodide/pyodide-webpack-example. You also have write permissions there, but it would still be good to make changes via PRs for that example repo.

I'm closing this issue as resolved, thanks @mneil! For follow-up discussions, it would probably be better to open separate issues in https://github.com/pyodide/pyodide-webpack-plugin.  I'll see if Pyodide repo has any relevant issues for webpack and if so transfer them there :)

@dlech thanks for starting this discussion. If you would like to collaborate on that plugin or test and provide feedback that would also be very welcome.
